### PR TITLE
add dask.array.cache

### LINKED ISSUE #293: da.cache
Should we provide a mechanism to cache dask.arrays?  This would compute and store the array and then create a new Array with a fresh graph.

```python
def cache(self, storage=None):
    """ Return new dask.array with empty graph that points to evaluated array """
    if storage is None:
        storage = ... # some temporary on-disk store
    self.store(storage)
    return da.from_array(storage, chunks=self.chunks)
```

This could be improved by the following:

1.  Figuring out how to do a nice temporary on-disk array store
2.  Supporting multiple inputs/stores at once

### PR DESCRIPTION
**Pull request by mrocklin**
This allows us to evaluate a dask.array, store in some cache,
and then return a new dask array pointing to that precomputed data.

One can specify a cache as either a MutableMapping or as an ndarray

Example
-------

```python
>>> import dask.array as da
>>> x = da.arange(5, chunks=2)
>>> y = 2*x + 1
>>> z = y.cache()  # triggers computation

>>> y.compute()  # Does entire computation
array([1, 3, 5, 7, 9])

>>> z.compute()  # Just pulls from store
array([1, 3, 5, 7, 9])

You might base a cache off of an array like a numpy array or
h5py.Dataset.

>>> cache = np.empty(5, dtype=x.dtype)
>>> z = y.cache(store=cache)
>>> cache
array([1, 3, 5, 7, 9])

Or one might use a MutableMapping like a dict or chest

>>> cache = dict()
>>> z = y.cache(store=cache)
>>> cache  # doctest: +SKIP
{('x', 0): array([1, 3]),
 ('x', 1): array([5, 7]),
 ('x', 2): array([9])}
```

Fixes #293

### CODE REVIEW

**Comment by shoyer** [comment]
Can we fail hard when trying to serialize an array with nbytes > 1 GB (or whatever the default Chest threshold is) if chest is not installed and you didn't provide a store? Or maybe better just to fail hard in general by default if chest is not installed? Defaulting to a plain dict is the sort of thing that could end up working poorly when called by third party code, and it's not much harder to write `.cache({})` if you don't need/want chest.

**Comment by mrocklin** [comment]
See recent commit

File: dask/array/core.py [hunk 1]
```diff
@@ -716,6 +716,70 @@ def compute(self, **kwargs):
...
+                if self.nbytes <= 1e9:
+                    store = dict()
+                else:
+                    raise ValueError("No out-of-core storage found."
+                        "Either:\n"
+                        "1. Install ``chest``, an out-of-core dictionary\n"
+                        "2. Provide an on-disk array like an h5py.Dataset") # pragma: no cover
+        if isinstance(store, MutableMapping):
+            name = next(names)
+            dsk = dict(((name, k[1:]), (operator.setitem, store, (tuple, list(k)), k))
```
Comment (shoyer) [inline dask/array/core.py]: Just for my edification -- why do you use `(tuple, list(k))` instead of just using `k` or `tuple(k)`?

Comment (mrocklin) [inline dask/array/core.py]: This stops `k` from being evaluated.  We don't want its value, we actually want the tuple itself.

`k` is a tuple like `('x', 0, 1)`.   This value is a key in our graph and so, unless we do something about it, `k` will be replaced with the full array rather than the tuple as we want.  `(tuple, list(k))` turns it into a list while in the graph but executes it back into a tuple when it's time to execute `setitem`.

That was complex, let me try again.

We want to do this

    store[('x', 0, 1)] = some_numpy_array

where `some_numpy_array` is the result corresponding to computing `('x', 0, 1)` from the graph.

Comment (shoyer) [inline dask/array/core.py]: Right, that makes sense. dask looks up tuples as keys in the dask.

It might be clearer to use a helper function for this with a descriptive name, e.g., `literal_value(k)` or `no_lookup(k)`.

Comment (mrocklin) [inline dask/array/core.py]: Yeah, that might make sense.  `quasiquote` might be a traditional name here.  Mind if I procrastinate on this?  This comes up in other contexts too and it might be good to handle them en masse

Comment (shoyer) [inline dask/array/core.py]: sure, that makes sense

File: dask/array/core.py [hunk 2]
```diff
@@ -716,6 +716,70 @@ def compute(self, **kwargs):
         result, = compute(self, **kwargs)
         return result
 
+    def cache(self, store=None, **kwargs):
```
Comment (shoyer) [inline dask/array/core.py]: document how `**kwargs` is used?

File: dask/array/tests/test_array_core.py [hunk 3]
```diff
@@ -1066,3 +1066,14 @@ def test_numpy_compat_is_notimplemented():
     x = da.from_array(a, chunks=5)
 
     assert_raises(NotImplementedError, add, x, a)
+
+
+def test_cache():
```
Comment (shoyer) [inline dask/array/tests/test_array_core.py]: Let's add a test for caching with an array store

Comment (mrocklin) [inline dask/array/tests/test_array_core.py]: Done

**Comment by shoyer** [comment]
LGTM :+1:
